[알트] 로또 2차 미션 코드리뷰 요청합니다.

KS-KIM opened the pull request:
### 리턴 타입이 `void`인 메서드를 어떻게 테스트 해야 좋을까요?

예를 들어, 로또 구입 금액을 가지고 있는 `PurchaseMoney`에서 `pay(long usingMoney)`라는 메서드를 호출하면 해당 금액을 가지고 있을 때 금액을 지불하고, 지불하지 못한다면 exception을 발생시킵니다.

이러한 상황에서는 getter 메서드를 두고 내부의 값이 어떻게 변했는지 확인하는 편이 좋을까요? 아니면 차라리 exception을 발생시키지 않고 지불의 성공 여부를 `boolean`값으로 반환하도록 만들어 테스트 코드에서 확인할 수 있도록 만드는 것이 좋을까요?

위와 같은 상황에서는 금액 지불의 성공 여부를 boolean값으로 리턴하여 어느정도 테스트상의 불편함을 해결할 수 있겠지만, 그렇지 않은 경우에는 `void` 리턴 타입을 가지는 메서드에 대해서 테스트 할 수 있는 좋은 방법이 있는지 궁금합니다!!

### 생성자는 어떤 방식으로 테스트 해야 좋을지 궁금합니다!

정상적인 생성자 호출의 경우
`assertThat(new Money(value)).isInstanceOf(Money.class)`

예외가 발생하는 생성자 호출의 경우 
`assertThatThrownBy(() -> new Money(value)).isInstanceOf(IllegalArgumentException.class)`

와 같은 형태로 테스트 코드를 작성하였는데, 어떤 값을 가지고  있는지는 확인하지 않기에 더 좋은 생성자 테스트 방법이 있는지 알고 싶습니다!!

### IDE에서 자동으로 생성할 수 있는 메서드도 모두 테스트의 대상일까요?

`toString()`, `equals()`, `hashCode()`, `get()`과 같이 직접 메서드를 정의하기 보다는 IDE의 자동 생성을 통해 만드는 메서드도 모두 테스트의 대상이 되겠죠...? 이전 미션에서는 직접 정의한 메서드 외에는 신경 쓰지 않았었는데, 자동 생성 메서드도 검증이 필요하다고 생각되어 질문 드립니다!!

KS-KIM committed:
refactor: 로또 금액을 외부로부터 받아 잔액에서 차감하도록 변경

KS-KIM committed:
feat: 수동 구매 구현

KS-KIM committed:
refactor: 변수명 수정

KS-KIM committed:
docs: 클래스 헤더 추가

KS-KIM committed:
feat: 구입 금액과 수동 구입 개수를 검증하기 위한 클래스 추가

KS-KIM committed:
refactor: 구입 정보를 입력받아 로또를 발급하는 LottoMachine 추가

KS-KIM committed:
test: 테스트 케이스 보완

기존에 테스트 하지 않았던 부분 (메서드, 예외 처리)에 대한 테스트 코드 추가

kingbbode commented on src/main/java/lotto/domain/LottoMachine.java:
```diff
@@ -0,0 +1,27 @@
+ */
+public class LottoMachine {
+	LottoGeneratable manualLottoTicketMachine;
+	LottoGeneratable autoLottoTicketMachine;
```
접근제어자가 private 이 아닌 이유가 있을까요?

KS-KIM replied:
접근제어자가 default일 이유가 없네요. private으로 수정하였습니다.

kingbbode commented on src/main/java/lotto/domain/LottoMachine.java:
```diff
@@ -0,0 +1,27 @@
+		LottoTicket manualLottoTicket = manualLottoTicketMachine.generate(purchaseMoney);
+		LottoTicket autoLottoTicket = autoLottoTicketMachine.generate(purchaseMoney);
+		return manualLottoTicket.concat(autoLottoTicket);
+	}
```
`List<String> lottoNumbers` 도 `PurchaseInformation` 에 포함되어야 하는 정보가 아닐까요?

kingbbode reviewed (changes requested):
아쉬운 부분이 남아 코멘트 작성해두었습니다 :)
확인 부탁드려요!

kingbbode commented:
- 리턴 타입이 void인 메서드를 어떻게 테스트 해야 좋을지
- 생성자는 어떤 방식으로 테스트 해야 좋을지
- IDE에서 자동으로 생성할 수 있는 메서드도 모두 테스트의 대상인지

를 질문해주셨는데요. 우리가 작성해야 하는 테스트는 작성되는 코드를 모두 검증하는 코드가 아니고, 의도한 설계한 테스트하는 코드입니다 :) 

- void 로 상태를 변화시키고, 그 상태를 개방하지 않는다면 그 상태를 사용하는 곳에서 테스트하면 될 것 같아요 :)
- 생성자는 말씀하신 방법으로 테스트하면 됩니다. 그 안에 값이 사용되는 부분에서 테스트될거에요.
- IDE에서 자동으로 생성할 수 있는 메서드는 의도한 바가 있다면 테스트하면 됩니다 :) 의도한 바가 없다면 꼭 할 필요는..?

KS-KIM committed:
style: static import문 wildcard 제거

KS-KIM committed:
docs: 기능 목록 업데이트

KS-KIM committed:
feat: 구입 개수를 가지는 PurchaseCount 추가

KS-KIM committed:
refactor: 로또 구입방식 개선

구매 정보를 가지는 PurchaseInformation 클래스를 제거하고, 구입 개수를 가지는 PurchaseCount를 추가했다.
PurchaseCount는 자동 로또 구매 개수를 구하기 위해 사용된다.
입력받은 구입액이 수동 구매 갯수만큼 구매하기에 부족한 경우 예외를 발생시키기 위함이다.

LottoMachine의 generate 메서드는 더이상 PurchaseInformation을 인자로 받지 않는다.
대신 인자로 금액을 받아 로또 구매 금액을 소모하며 티켓을 생성한다.

KS-KIM committed:
test: 로또 수 생성 로직 테스트 추가

KS-KIM committed:
fix: 메시지 출력 순서 변경

KS-KIM commented:
> `List<String> lottoNumbers` 도 `PurchaseInformation` 에 포함되어야 하는 정보가 아닐까요?

`PurchaseInformation`을 만든 목적은 구입 금액이 수동 구매 개수만큼 구입하기에 모자르는 경우 수동 로또 숫자를 입력받기 전에 사전 검증하여 exception을 던지기 위함이었습니다. 이 객체를 `LottoMachine`에 넘길 필요가 있을까 고민해보니 굳이 그럴 필요가 없다는 것을 깨달았습니다. `LottoMachine`은 구입에 대한 정보를 모르더라도 `PurchaseMoney`만 받으면 모든 금액을 사용하여 수동과 자동 로또를 구입하도록 설계되었기 때문입니다.

이러한 점을 인지하고 리팩토링 과정에서는 `LottoMachine` 또한 `ManualLottoTicketFactory`, `AutoLottoTicketFactory`와 마찬가지로 `LottoGenerative`의 concrete class로 만들어 `generate(PurchaseMoney): LottoTicket` 메서드를 오버라이드 하도록 변경했습니다.

`PurchaseInformation`은 역할이 모호해져 삭제하였습니다. 하지만 여전히 수동 로또와 자동 로또의 개수를 출력해야 했기에 구입 개수를 감싸는 원시객체 `PurchaseCount`를 추가하였습니다.

---

로또 생성을 포괄하는 `LottoMachine`에서 로또 생성 객체를 주입받을 것인가, 아니면 클래스 내에서 생성자 호출을 통해 새로운 객체를 생성할 것인가에 대해서는 조금 더 고민해 보아야 할 것 같습니다. 외부로부터 로또 생성 객체를 주입받는다면 요구사항이 변경되어 또 다른 로또 생성 방식이 추가되었을 때 유연하게 대응할 수 있겠지만, 구입 금액을 받아 처리하는 부분에 변화가 필요할 것 같습니다. 기존의 처리 방식은 수동 로또를 구매하고 남은 금액을 모두 자동으로 구매하는 방식이기 때문에 로또 생성 객체를 호출하는 순서나 구입을 처리하는 방식에 따라 결과가 달라질 수 있기 때문입니다. 만약 외부에서 로또 생성 객체를 주입받는 방식을 채택하게 된다면 각 생성 방식별로 투입하는 금액을 다르게 처리하는 방법이 있을 것 같습니다.

위 고민은 '로또 게임에 요구사항이 더 추가된다면?'이라는 고민에서 시작한 것인데, 일어나지 않을 일까지 과하게 생각한건 아닌가 싶습니다. 그래서 2단계 로또 미션에서는 '수동 로또 기능 추가'라는 목표에 집중했습니다. '확장 가능성을 열어두고 최대한 변화에 자유롭게 설계하기'와 '주어진 요구사항에 대응하는 빠른 설계하기' 사이에서 어떤것을 택해야 할지는 경험을 쌓으며 배워나가겠습니다!! :)

kingbbode commented on src/main/java/lotto/domain/LottoMachine.java:
```diff
@@ -9,17 +9,16 @@
-public class LottoMachine {
-	LottoGeneratable manualLottoTicketMachine;
-	LottoGeneratable autoLottoTicketMachine;
+public class LottoMachine implements LottoGenerative {
```
`LottoMachine` 보다는 다른 구현체와 일관성을 가질 수 있는 네이밍이였으면 더 좋았겠네요 :)

kingbbode replied:
구현하신 방식은 `delegate pattern` 이라는 방식 중 하나입니다. 좋은 선택이라고 생각됩니다 :)

kingbbode commented on src/main/java/lotto/domain/LottoMachine.java:
```diff
@@ -9,17 +9,16 @@
-	LottoGeneratable autoLottoTicketMachine;
+public class LottoMachine implements LottoGenerative {
+	private LottoGenerative manualLottoTicketMachine;
+	private LottoGenerative autoLottoTicketMachine;
```
List 로 관리한다면 앞으로의 변경에 조금 더 유연해질 수 있지 않을까요?

kingbbode commented on src/main/java/lotto/domain/LottoMachine.java:
```diff
@@ -9,17 +9,16 @@
+	private LottoGenerative manualLottoTicketMachine;
+	private LottoGenerative autoLottoTicketMachine;
 
 	public LottoMachine(List<String> lottoNumbers) {
```
생성자 인자와 구성 필드가 다르게 존재할 때는 “정적 팩토리 메서드”를 사용하는 것이 더 의도를 드러낼 수 있는 방법일 것 같습니다 :)

kingbbode reviewed (approved):
고민의 생각과 방향들이 좋네요 :)
앞으로도 기대하겠습니다. 👍 
이만 머지하겠습니다.
